File: go/verifier/parse.go
package verifier

import (
	"regexp"
	"strings"
)

var currencyRe = regexp.MustCompile(`^[A-Z]{3}$`)
var jwsRe = regexp.MustCompile(`^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+$`)

func nonEmpty(v any) (string, bool) {
	s, ok := v.(string)
	if !ok || strings.TrimSpace(s) == "" || len(s) > 512 {
		return "", false
	}
	return strings.TrimSpace(s), true
}

// ParseVerifyRequest validates an untrusted request; the second value is the failing field ("" when ok).
func ParseVerifyRequest(o map[string]any) (VerifyRequest, string) {
	var r VerifyRequest
	if o == nil {
		return r, "not_object"
	}
	var ok bool
	if r.Agent, ok = nonEmpty(o["agent"]); !ok {
		return r, "agent"
	}
	if r.Action, ok = nonEmpty(o["action"]); !ok {
		return r, "action"
	}
	for _, f := range []struct {
		key string
		dst **string
	}{{"resource", &r.Resource}, {"principal", &r.Principal}, {"delegation_id", &r.DelegationID}} {
		if v, present := o[f.key]; present && v != nil {
			s, ok := nonEmpty(v)
			if !ok {
				return r, f.key
			}
			*f.dst = strp(s)
		}
	}
	if v, present := o["amount"]; present && v != nil {
		f, ok := asFloat(v)
		if !ok || f < 0 {
			return r, "amount"
		}
		r.Amount = &f
	}
	if v, present := o["currency"]; present && v != nil {
		s, ok := v.(string)
		if !ok || !currencyRe.MatchString(s) {
			return r, "currency"
		}
		r.Currency = strp(s)
	}
	if v, present := o["attestation"]; present && v != nil {
		s, ok := v.(string)
		if !ok || !jwsRe.MatchString(s) || len(s) > 16384 {
			return r, "attestation"
		}
		r.Attestation = strp(s)
	}
	if v, present := o["context"]; present && v != nil {
		m, ok := v.(map[string]any)
		if !ok {
			return r, "context"
		}
		r.Context = m
	}
	return r, ""
}

File: packages/verifier/src/parse.ts
import type { VerifyRequest } from "./types";

export type ParseResult =
  | { ok: true; request: VerifyRequest }
  | { ok: false; code: "not_object" | "agent" | "action" | "resource" | "principal" | "delegation_id" | "amount" | "currency" | "context" | "attestation" };

const nonEmpty = (v: unknown): v is string => typeof v === "string" && v.trim().length > 0 && v.length <= 512;
const optStr = (v: unknown): boolean => v === undefined || nonEmpty(v);

/** Validate an untrusted verify request. Never throws; unknown keys are dropped. */
export function parseVerifyRequest(input: unknown): ParseResult {
  if (input === null || typeof input !== "object" || Array.isArray(input)) return { ok: false, code: "not_object" };
  const o = input as Record<string, unknown>;
  if (!nonEmpty(o.agent)) return { ok: false, code: "agent" };
  if (!nonEmpty(o.action)) return { ok: false, code: "action" };
  if (!optStr(o.resource)) return { ok: false, code: "resource" };
  if (!optStr(o.principal)) return { ok: false, code: "principal" };
  if (!optStr(o.delegation_id)) return { ok: false, code: "delegation_id" };
  if (o.amount !== undefined && !(typeof o.amount === "number" && Number.isFinite(o.amount) && o.amount >= 0)) return { ok: false, code: "amount" };
  if (o.currency !== undefined && !(typeof o.currency === "string" && /^[A-Z]{3}$/.test(o.currency))) return { ok: false, code: "currency" };
  if (o.attestation !== undefined && !(typeof o.attestation === "string" && /^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+$/.test(o.attestation) && o.attestation.length <= 16384)) return { ok: false, code: "attestation" };
  if (o.context !== undefined && (o.context === null || typeof o.context !== "object" || Array.isArray(o.context))) return { ok: false, code: "context" };
  const request: VerifyRequest = { agent: (o.agent as string).trim(), action: (o.action as string).trim() };
  if (o.resource !== undefined) request.resource = (o.resource as string).trim();
  if (o.principal !== undefined) request.principal = (o.principal as string).trim();
  if (o.delegation_id !== undefined) request.delegation_id = (o.delegation_id as string).trim();
  if (o.amount !== undefined) request.amount = o.amount as number;
  if (o.currency !== undefined) request.currency = o.currency as string;
  if (o.context !== undefined) request.context = o.context as Record<string, unknown>;
  if (o.attestation !== undefined) request.attestation = o.attestation as string;
  return { ok: true, request };
}
